cfgov/tccp/models.py

```diff
@@ -40,109 +40,289 @@ def available_in(self, value):
-                    f"{basename}_apr_{tier_column_suffix}"
+        # 1. If the column "targeted_credit_tiers" column doesn't include the
+        #    good tier, assign the card a purchase APR of None, which excludes
+        #    it from aggregation calculations.
+        # 2. Otherwise, if the column "purchase_apr_good" is defined, use it
+        #    as the card's purchase APR.
+        # 3. Otherwise, if the column "purchase_apr_median" is defined, use it
+        #    as the card's purchase APR.
+        # 4. Otherwise, if both columns "purchase_apr_min" and
+        #    "purchase_apr_max" are defined, use the the max APR as the card's
+        #    purchase APR.
+        # 5. Otherwise, use an APR of None, which excludes the card from
+        #    aggregation calculations.
+        #
+        # This logic can be represented in SQL roughly as follows to aggregate
+        # the purchase_apr_good_pct25 statistic:
+        #
+        # SELECT
+        #   PERCENTILE_CONT(0.25)
+        #   WITHIN GROUP (
+        #     ORDER BY CASE
+        #       WHEN
+        #         targeted_credit_tiers" @> ["Credit scores from 620 to 719"]
+        #       THEN
+        #         COALESCE(
+        #           purchase_apr_good,
+        #           purchase_apr_median,
+        #           CASE
+        #             WHEN
+        #               purchase_apr_min IS NOT NULL AND
+        #               purchase_apr_max IS NOT NULL
+        #             THEN
+        #               purchase_apr_max
+        #             ELSE NULL
+        #           END
+        #         )
+        #       ELSE NULL
+        #     END
+        #   ) AS purchase_apr_good_pct25
+        # FROM ...
+        #
+        # This logic is expressed in Django code here:
+        aggregates = {
+            f"purchase_apr_{tier_suffix}_{stat_name}": stat_fn(
+                Case(
+                    When(
+                        targeted_credit_tiers__contains=tier_value,
+                        then=Coalesce(
+                            F(f"purchase_apr_{tier_suffix}"),
+                            F("purchase_apr_median"),
+                            Case(
+                                When(
+                                    Q(purchase_apr_min__isnull=False)
+                                    & Q(purchase_apr_max__isnull=False),
+                                    then=F("purchase_apr_max"),
+                                ),
+                            ),
+                        ),
+                    ),
                 )
-                for basename in ("purchase", "transfer")
+            )
+            for (stat_name, stat_fn), (
+                tier_value,
+                tier_suffix,
+            ) in product(stats, enums.CreditTierColumns)
+        }
+
+        aggregates.update(
+            {
+                "count": Count("pk"),
+                "first_report_date": Min("report_date"),
             }
         )
```

Comment:
Popping into an old PR to express my appreciation and admiration. This is really, really neat. And excellently explained in comment ❤️